### ISSUE TITLE
Zero page start can't be adjusted

### ISSUE THREAD

**Issue by SvOlli** [SamCoVT/TaliForth2#139]
Hello!

On my Sorbus Computer, these changes break TaliForth2 in a way that it compiles, but cannot be executed. It runs in some kind of BRKs throwing execution off rails.

Build binary crashes (2024-06-05)

- 283f50783a03307399c240a131d889e8896f31d8 slim down other platform config, default kbhit implementation
- a79de57321d40fedafc1a3ab4dd554cf5f72a5d9 revert binary
- f0ee88385371d059e79b51bb7733307d5166c6c4 simplify configuration, update README
- 4d3197ccdff86416ecd54b770236c5b6de0ad21a adds c65 platform and key? support
- 447531f3ee29d2f4dbb916167af51559dc199953 wip

Build binary works (2024-05-31)

I couldn't dissect this further, because in between the code does not compile.

While writing this up, I could figure out the problem: the Sorbus Computer has some reserved zero page addresses at $0000-$000f. This can't be changed in the config file.

What is the best way to fix this? Have ram_start be defined as `?=` instead of `=`, so it can be overwritten in the config?

**Comment by SamCoVT** [comment]
All of the variables defined in zero page should be using the starting label `user0`, which it looks like you have properly overridden in your platform file.

I will recommend upgrading to a newer/latest version of 64tass.  The version that was shipping with multiple Linux distros (1.56.2625) has a bug that can result in Tali compiling, but not running properly.  I'm currently using v1.59.3120 now.

I'll take a look at it this evening to make sure the zero page variables are being located correctly.  Is the platform-sorbus.asm file in the repository still good to use?

**Comment by SvOlli** [comment]
Take a look at definitions.asm, line 18. There is a hard coded ram_start = $0000.

**Comment by SvOlli** [comment]
Setting it to $0010 makes it work again with an unchanged platform-sorbus.asm.

**Comment by SamCoVT** [comment]
That would seem to indicate that some piece of code is using `ram_start` (or `zpage`).  It shouldn't, because `user0` is supposed to be the first location in zero page that Tali uses.
Further investigation shows the culprit is in COLD (xt_cold) on line 55 of words/all.asm where it says:
```
_load_zp_loop:
                ; This loop loads them back to front.
                lda cold_zp_table,y
                sta zpage,y
```
That should be:
```
_load_zp_loop:
                ; This loop loads them back to front.
                lda cold_zp_table,y
                sta user0,y
```
That should make it correspond with the `cold_zp_table` in definitions.asm, which is expecting that data to start at whatever address `user0` holds.  That would certainly explain why Tali isn't starting up properly.

If you change that sta instruction, I believe your `user0` value from your platform file should work.
It's probably worth taking a look at and simplifying the labels that are used to specify where Tali puts things.  The ram_start and zpage labels describe the addresses used by the 65C02 hardware (regardless of whether there is I/O in zero page) and aren't (supposed to be) used except to set up the default value of user0.  We could certainly simplify this all down to a single set of labels that the user can override in their platform files, with updated documentation for those in the sample platform files.

Let me know if that code change works for you.  You're welcome to submit a pull request or I can fix it.  I'll leave this issue open as the documentation should be updated regardless of whether or not the labels are cleaned up.

**Comment by SvOlli** [comment]
Yes, it works. Pull request created.

**Comment by SamCoVT** [comment]
Merged PR #140
